File: go/schema.go
package amino

import (
	"fmt"
	"regexp"
	"sort"
	"strconv"
	"strings"
)

// Kind classifies a field's declared type.
type Kind int

const (
	KindInt Kind = iota
	KindFloat
	KindStr
	KindBool
	KindList
	KindCustom // a struct name or a registered custom type
)

// Field is one declared field, at the top level or inside a struct.
type Field struct {
	Name         string
	Kind         Kind
	TypeName     string         // "Int", "List[Str]", "Address", "email"
	ElementTypes []string       // for lists
	Constraints  map[string]any // {min: 300, oneOf: [...]}
	Optional     bool
}

// Struct is a named group of fields usable as a type.
type Struct struct {
	Name   string
	Fields []Field
}

// Param is one declared function parameter.
type Param struct {
	Name     string
	TypeName string
	Optional bool
}

// Function is a declared function signature. The implementation is supplied
// with WithFunction.
type Function struct {
	Name       string
	Params     []Param
	ReturnType string
}

// Schema is a parsed and validated .amn document.
type Schema struct {
	Fields    []Field
	Structs   []Struct
	Functions []Function

	structs map[string]*Struct
	paths   map[string]*Field // "customer.billing.city" -> field
	funcs   map[string]*Function
}

var primitives = map[string]Kind{"Int": KindInt, "Float": KindFloat, "Str": KindStr, "Bool": KindBool}

var reserved = map[string]bool{"struct": true, "List": true}

// ParseSchema parses .amn text. Validation against custom types happens in
// LoadSchema, which knows which custom types are registered.
func ParseSchema(text string) (*Schema, error) {
	p := &schemaParser{text: text, line: 1}
	return p.parse()
}

type schemaParser struct {
	text string
	pos  int
	line int
}

func (p *schemaParser) peek() (byte, bool) {
	if p.pos < len(p.text) {
		return p.text[p.pos], true
	}
	return 0, false
}

func (p *schemaParser) advance() byte {
	ch := p.text[p.pos]
	if ch == '\n' {
		p.line++
	}
	p.pos++
	return ch
}

func (p *schemaParser) skipWS(newlines bool) {
	for p.pos < len(p.text) {
		ch := p.text[p.pos]
		switch {
		case ch == '#':
			for p.pos < len(p.text) && p.text[p.pos] != '\n' {
				p.pos++
			}
		case ch == ' ' || ch == '\t':
			p.pos++
		case newlines && (ch == '\n' || ch == '\r'):
			if ch == '\n' {
				p.line++
			}
			p.pos++
		default:
			return
		}
	}
}

func (p *schemaParser) skipH() {
	for p.pos < len(p.text) && (p.text[p.pos] == ' ' || p.text[p.pos] == '\t') {
		p.pos++
	}
}

var identRe = regexp.MustCompile(`^[a-zA-Z_][a-zA-Z0-9_]*`)

func (p *schemaParser) readIdent() (string, error) {
	m := identRe.FindString(p.text[p.pos:])
	if m == "" {
		return "", newError(CodeSchemaParse, "expected identifier at line %d", p.line)
	}
	p.pos += len(m)
	return m, nil
}

func (p *schemaParser) expect(ch byte) error {
	p.skipH()
	got, ok := p.peek()
	if !ok || got != ch {
		return newError(CodeSchemaParse, "expected %q at line %d", string(ch), p.line)
	}
	p.advance()
	return nil
}

func (p *schemaParser) parseStrLiteral() (string, error) {
	p.advance() // opening '
	start := p.pos
	for {
		ch, ok := p.peek()
		if !ok {
			return "", newError(CodeSchemaParse, "unterminated string at line %d", p.line)
		}
		if ch == '\'' {
			break
		}
		p.advance()
	}
	s := p.text[start:p.pos]
	p.advance()
	return s, nil
}

func (p *schemaParser) parseListLit() ([]any, error) {
	p.advance() // [
	items := []any{}
	p.skipH()
	for {
		ch, ok := p.peek()
		if !ok {
			return nil, newError(CodeSchemaParse, "unterminated list at line %d", p.line)
		}
		if ch == ']' {
			break
		}
		v, err := p.parseConstraintVal()
		if err != nil {
			return nil, err
		}
		items = append(items, v)
		p.skipH()
		if ch, _ := p.peek(); ch == ',' {
			p.advance()
			p.skipH()
		}
	}
	p.advance() // ]
	return items, nil
}

var (
	floatRe = regexp.MustCompile(`^-?\d+\.\d+`)
	intRe   = regexp.MustCompile(`^-?\d+`)
	boolRe  = regexp.MustCompile(`^(true|false)`)
)

func (p *schemaParser) parseConstraintVal() (any, error) {
	p.skipH()
	ch, ok := p.peek()
	if !ok {
		return nil, newError(CodeSchemaParse, "expected constraint value at line %d", p.line)
	}
	if ch == '\'' {
		return p.parseStrLiteral()
	}
	if ch == '[' {
		return p.parseListLit()
	}
	rest := p.text[p.pos:]
	if m := floatRe.FindString(rest); m != "" {
		p.pos += len(m)
		f, _ := strconv.ParseFloat(m, 64)
		return f, nil
	}
	if m := intRe.FindString(rest); m != "" {
		p.pos += len(m)
		i, _ := strconv.ParseInt(m, 10, 64)
		return i, nil
	}
	if m := boolRe.FindString(rest); m != "" {
		p.pos += len(m)
		return m == "true", nil
	}
	return nil, newError(CodeSchemaParse, "expected constraint value at line %d", p.line)
}

func (p *schemaParser) parseConstraints() (map[string]any, error) {
	p.advance() // {
	out := map[string]any{}
	p.skipH()
	for {
		ch, ok := p.peek()
		if !ok {
			return nil, newError(CodeSchemaParse, "unterminated constraint block at line %d", p.line)
		}
		if ch == '}' {
			break
		}
		key, err := p.readIdent()
		if err != nil {
			return nil, err
		}
		if err := p.expect(':'); err != nil {
			return nil, err
		}
		v, err := p.parseConstraintVal()
		if err != nil {
			return nil, err
		}
		out[key] = v
		p.skipH()
		if ch, _ := p.peek(); ch == ',' {
			p.advance()
			p.skipH()
		}
	}
	p.advance() // }
	return out, nil
}

func (p *schemaParser) parseTypeExpr() (Kind, string, []string, error) {
	p.skipH()
	name, err := p.readIdent()
	if err != nil {
		return 0, "", nil, err
	}
	if name == "List" {
		if err := p.expect('['); err != nil {
			return 0, "", nil, err
		}
		first, err := p.readIdent()
		if err != nil {
			return 0, "", nil, err
		}
		elems := []string{first}
		for {
			ch, _ := p.peek()
			if ch != '|' {
				break
			}
			p.advance()
			e, err := p.readIdent()
			if err != nil {
				return 0, "", nil, err
			}
			elems = append(elems, e)
		}
		if err := p.expect(']'); err != nil {
			return 0, "", nil, err
		}
		return KindList, "List[" + strings.Join(elems, "|") + "]", elems, nil
	}
	if k, ok := primitives[name]; ok {
		return k, name, nil, nil
	}
	return KindCustom, name, nil, nil
}

func (p *schemaParser) parseField() (Field, error) {
	p.skipH()
	name, err := p.readIdent()
	if err != nil {
		return Field{}, err
	}
	if reserved[name] {
		return Field{}, newError(CodeSchemaParse, "reserved word %q used as field name at line %d", name, p.line)
	}
	if err := p.expect(':'); err != nil {
		return Field{}, err
	}
	kind, tname, elems, err := p.parseTypeExpr()
	if err != nil {
		return Field{}, err
	}
	f := Field{Name: name, Kind: kind, TypeName: tname, ElementTypes: elems, Constraints: map[string]any{}}
	p.skipH()
	if ch, _ := p.peek(); ch == '?' {
		f.Optional = true
		p.advance()
	}
	p.skipH()
	if ch, _ := p.peek(); ch == '{' {
		c, err := p.parseConstraints()
		if err != nil {
			return Field{}, err
		}
		f.Constraints = c
	}
	return f, nil
}

func (p *schemaParser) parseStruct() (Struct, error) {
	if _, err := p.readIdent(); err != nil { // 'struct'
		return Struct{}, err
	}
	p.skipH()
	name, err := p.readIdent()
	if err != nil {
		return Struct{}, err
	}
	p.skipWS(true)
	if err := p.expect('{'); err != nil {
		return Struct{}, err
	}
	s := Struct{Name: name}
	for {
		p.skipWS(true)
		ch, ok := p.peek()
		if !ok {
			return Struct{}, newError(CodeSchemaParse, "unterminated struct %q", name)
		}
		if ch == '}' {
			break
		}
		f, err := p.parseField()
		if err != nil {
			return Struct{}, err
		}
		s.Fields = append(s.Fields, f)
		p.skipH()
		if ch, _ := p.peek(); ch == ',' {
			p.advance()
		}
	}
	p.advance() // }
	return s, nil
}

func (p *schemaParser) isFunction() bool {
	savedPos, savedLine := p.pos, p.line
	defer func() { p.pos, p.line = savedPos, savedLine }()
	if _, err := p.readIdent(); err != nil {
		return false
	}
	p.skipH()
	if ch, _ := p.peek(); ch != ':' {
		return false
	}
	p.advance()
	p.skipH()
	ch, _ := p.peek()
	return ch == '('
}

func (p *schemaParser) parseFunction() (Function, error) {
	name, err := p.readIdent()
	if err != nil {
		return Function{}, err
	}
	if err := p.expect(':'); err != nil {
		return Function{}, err
	}
	if err := p.expect('('); err != nil {
		return Function{}, err
	}
	fn := Function{Name: name}
	p.skipH()
	for {
		ch, ok := p.peek()
		if !ok {
			return Function{}, newError(CodeSchemaParse, "unterminated parameter list at line %d", p.line)
		}
		if ch == ')' {
			break
		}
		pname, err := p.readIdent()
		if err != nil {
			return Function{}, err
		}
		if err := p.expect(':'); err != nil {
			return Function{}, err
		}
		_, tname, _, err := p.parseTypeExpr()
		if err != nil {
			return Function{}, err
		}
		param := Param{Name: pname, TypeName: tname}
		p.skipH()
		if ch, _ := p.peek(); ch == '?' {
			param.Optional = true
			p.advance()
		}
		fn.Params = append(fn.Params, param)
		p.skipH()
		if ch, _ := p.peek(); ch == ',' {
			p.advance()
			p.skipH()
		}
	}
	p.advance() // )
	p.skipH()
	if !strings.HasPrefix(p.text[p.pos:], "->") {
		return Function{}, newError(CodeSchemaParse, "expected '->' at line %d", p.line)
	}
	p.pos += 2
	p.skipH()
	ret, err := p.readIdent()
	if err != nil {
		return Function{}, err
	}
	p.skipH()
	if ch, _ := p.peek(); ch == '?' {
		p.advance()
	}
	fn.ReturnType = ret
	return fn, nil
}

var structKeywordRe = regexp.MustCompile(`^struct(?:[^a-zA-Z0-9_]|$)`)

func (p *schemaParser) parse() (*Schema, error) {
	s := &Schema{}
	for {
		p.skipWS(true)
		if p.pos >= len(p.text) {
			break
		}
		switch {
		case structKeywordRe.MatchString(p.text[p.pos:]):
			st, err := p.parseStruct()
			if err != nil {
				return nil, err
			}
			s.Structs = append(s.Structs, st)
		case p.isFunction():
			fn, err := p.parseFunction()
			if err != nil {
				return nil, err
			}
			s.Functions = append(s.Functions, fn)
		default:
			f, err := p.parseField()
			if err != nil {
				return nil, err
			}
			s.Fields = append(s.Fields, f)
		}
	}
	return s, nil
}

// validate checks references, duplicates, and cycles, then builds the lookup
// indexes. customTypes are the registered custom type names.
func (s *Schema) validate(customTypes map[string]bool) error {
	s.structs = map[string]*Struct{}
	for i := range s.Structs {
		s.structs[s.Structs[i].Name] = &s.Structs[i]
	}
	known := map[string]bool{"Int": true, "Float": true, "Str": true, "Bool": true, "List": true}
	for name := range s.structs {
		known[name] = true
	}
	for name := range customTypes {
		known[name] = true
	}

	seen := map[string]bool{}
	check := func(name string) error {
		if seen[name] {
			return newError(CodeSchemaValidation, "duplicate name %q", name)
		}
		seen[name] = true
		return nil
	}
	for _, f := range s.Fields {
		if err := check(f.Name); err != nil {
			return err
		}
	}
	for _, st := range s.Structs {
		if err := check(st.Name); err != nil {
			return err
		}
	}
	for _, fn := range s.Functions {
		if err := check(fn.Name); err != nil {
			return err
		}
	}

	for _, f := range s.Fields {
		if f.Kind == KindCustom && !known[f.TypeName] {
			return fieldError(CodeSchemaValidation, f.Name, "unknown type %q", f.TypeName)
		}
	}
	for _, st := range s.Structs {
		fieldSeen := map[string]bool{}
		for _, f := range st.Fields {
			if fieldSeen[f.Name] {
				return newError(CodeSchemaValidation, "duplicate field %q in struct %q", f.Name, st.Name)
			}
			fieldSeen[f.Name] = true
			if f.Kind == KindCustom && !known[f.TypeName] {
				return newError(CodeSchemaValidation, "unknown type %q in struct %q", f.TypeName, st.Name)
			}
		}
	}

	var dfs func(name string, visiting map[string]bool) error
	dfs = func(name string, visiting map[string]bool) error {
		if visiting[name] {
			return newError(CodeSchemaValidation, "circular struct reference involving %q", name)
		}
		st, ok := s.structs[name]
		if !ok {
			return nil
		}
		next := map[string]bool{name: true}
		for k := range visiting {
			next[k] = true
		}
		for _, f := range st.Fields {
			if f.Kind == KindCustom {
				if _, isStruct := s.structs[f.TypeName]; isStruct {
					if err := dfs(f.TypeName, next); err != nil {
						return err
					}
				}
			}
		}
		return nil
	}
	names := make([]string, 0, len(s.structs))
	for name := range s.structs {
		names = append(names, name)
	}
	sort.Strings(names)
	for _, name := range names {
		if err := dfs(name, map[string]bool{}); err != nil {
			return err
		}
	}

	s.paths = map[string]*Field{}
	for i := range s.Fields {
		f := &s.Fields[i]
		s.paths[f.Name] = f
		if _, isStruct := s.structs[f.TypeName]; isStruct {
			s.indexStruct(f.Name, f.TypeName)
		}
	}
	s.funcs = map[string]*Function{}
	for i := range s.Functions {
		s.funcs[s.Functions[i].Name] = &s.Functions[i]
	}
	return nil
}

func (s *Schema) indexStruct(prefix, structName string) {
	st, ok := s.structs[structName]
	if !ok {
		return
	}
	for i := range st.Fields {
		f := &st.Fields[i]
		key := prefix + "." + f.Name
		s.paths[key] = f
		if _, isStruct := s.structs[f.TypeName]; isStruct {
			s.indexStruct(key, f.TypeName)
		}
	}
}

// Lookup returns the field at a dotted path, or nil.
func (s *Schema) Lookup(path string) *Field { return s.paths[path] }

// LookupFunction returns the declared function, or nil.
func (s *Schema) LookupFunction(name string) *Function { return s.funcs[name] }

// IsStruct reports whether name is a declared struct.
func (s *Schema) IsStruct(name string) bool { _, ok := s.structs[name]; return ok }

// Export renders the schema back to .amn text: structs, then fields, then functions.
func (s *Schema) Export() string {
	var b strings.Builder
	fieldStr := func(f Field) string {
		q := ""
		if f.Optional {
			q = "?"
		}
		c := ""
		if len(f.Constraints) > 0 {
			keys := make([]string, 0, len(f.Constraints))
			for k := range f.Constraints {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			parts := make([]string, 0, len(keys))
			for _, k := range keys {
				parts = append(parts, fmt.Sprintf("%s: %s", k, exportValue(f.Constraints[k])))
			}
			c = " {" + strings.Join(parts, ", ") + "}"
		}
		return f.Name + ": " + f.TypeName + q + c
	}
	for _, st := range s.Structs {
		parts := make([]string, 0, len(st.Fields))
		for _, f := range st.Fields {
			parts = append(parts, fieldStr(f))
		}
		fmt.Fprintf(&b, "struct %s {%s}\n", st.Name, strings.Join(parts, ", "))
	}
	for _, f := range s.Fields {
		b.WriteString(fieldStr(f))
		b.WriteByte('\n')
	}
	for _, fn := range s.Functions {
		params := make([]string, 0, len(fn.Params))
		for _, p := range fn.Params {
			q := ""
			if p.Optional {
				q = "?"
			}
			params = append(params, p.Name+": "+p.TypeName+q)
		}
		fmt.Fprintf(&b, "%s: (%s) -> %s\n", fn.Name, strings.Join(params, ", "), fn.ReturnType)
	}
	return strings.TrimRight(b.String(), "\n")
}

func exportValue(v any) string {
	switch x := v.(type) {
	case string:
		return "'" + x + "'"
	case []any:
		parts := make([]string, len(x))
		for i, e := range x {
			parts[i] = exportValue(e)
		}
		return "[" + strings.Join(parts, ", ") + "]"
	case bool:
		if x {
			return "true"
		}
		return "false"
	default:
		return fmt.Sprint(x)
	}
}

File: typescript/src/schema.ts
import { AminoError } from "./errors.js";

export type Kind = "Int" | "Float" | "Str" | "Bool" | "List" | "Custom";

export interface Field {
  name: string;
  kind: Kind;
  /** "Int", "List[Str]", "Address", "email" */
  typeName: string;
  elementTypes: string[];
  constraints: Record<string, unknown>;
  optional: boolean;
}

export interface StructDef {
  name: string;
  fields: Field[];
}

export interface Param {
  name: string;
  typeName: string;
  optional: boolean;
}

export interface FunctionDef {
  name: string;
  params: Param[];
  returnType: string;
}

const PRIMITIVES: Record<string, Kind> = { Int: "Int", Float: "Float", Str: "Str", Bool: "Bool" };
const RESERVED = new Set(["struct", "List"]);
const IDENT = /^[a-zA-Z_][a-zA-Z0-9_]*/;
const FLOAT = /^-?\d+\.\d+/;
const INT = /^-?\d+/;
const BOOL = /^(true|false)/;
const STRUCT_KW = /^struct(?![a-zA-Z0-9_])/;

/** A parsed and validated .amn document. */
export class Schema {
  readonly fields: Field[] = [];
  readonly structs: StructDef[] = [];
  readonly functions: FunctionDef[] = [];

  private structMap = new Map<string, StructDef>();
  private paths = new Map<string, Field>();
  private funcMap = new Map<string, FunctionDef>();

  /** Field at a dotted path, or undefined. */
  lookup(path: string): Field | undefined {
    return this.paths.get(path);
  }

  lookupFunction(name: string): FunctionDef | undefined {
    return this.funcMap.get(name);
  }

  isStruct(name: string): boolean {
    return this.structMap.has(name);
  }

  /** Checks references, duplicates, and cycles, then builds lookup indexes. */
  validate(customTypes: Set<string>): void {
    this.structMap = new Map(this.structs.map((s) => [s.name, s]));
    const known = new Set(["Int", "Float", "Str", "Bool", "List", ...this.structMap.keys(), ...customTypes]);

    const seen = new Set<string>();
    for (const item of [...this.fields, ...this.structs, ...this.functions]) {
      if (seen.has(item.name)) throw new AminoError("schema_validation", `duplicate name "${item.name}"`);
      seen.add(item.name);
    }
    for (const f of this.fields) {
      if (f.kind === "Custom" && !known.has(f.typeName)) {
        throw new AminoError("schema_validation", `unknown type "${f.typeName}"`, f.name);
      }
    }
    for (const s of this.structs) {
      const fieldSeen = new Set<string>();
      for (const f of s.fields) {
        if (fieldSeen.has(f.name)) {
          throw new AminoError("schema_validation", `duplicate field "${f.name}" in struct "${s.name}"`);
        }
        fieldSeen.add(f.name);
        if (f.kind === "Custom" && !known.has(f.typeName)) {
          throw new AminoError("schema_validation", `unknown type "${f.typeName}" in struct "${s.name}"`);
        }
      }
    }

    const dfs = (name: string, visiting: Set<string>): void => {
      if (visiting.has(name)) throw new AminoError("schema_validation", `circular struct reference involving "${name}"`);
      const s = this.structMap.get(name);
      if (!s) return;
      const next = new Set(visiting).add(name);
      for (const f of s.fields) {
        if (f.kind === "Custom" && this.structMap.has(f.typeName)) dfs(f.typeName, next);
      }
    };
    for (const name of [...this.structMap.keys()].sort()) dfs(name, new Set());

    this.paths = new Map();
    for (const f of this.fields) {
      this.paths.set(f.name, f);
      if (this.structMap.has(f.typeName)) this.indexStruct(f.name, f.typeName);
    }
    this.funcMap = new Map(this.functions.map((fn) => [fn.name, fn]));
  }

  private indexStruct(prefix: string, structName: string): void {
    const s = this.structMap.get(structName);
    if (!s) return;
    for (const f of s.fields) {
      const key = `${prefix}.${f.name}`;
      this.paths.set(key, f);
      if (this.structMap.has(f.typeName)) this.indexStruct(key, f.typeName);
    }
  }

  /** Renders the schema as .amn text: structs, then fields, then functions. */
  export(): string {
    const fieldStr = (f: Field): string => {
      const q = f.optional ? "?" : "";
      const keys = Object.keys(f.constraints).sort();
      const c = keys.length ? ` {${keys.map((k) => `${k}: ${exportValue(f.constraints[k])}`).join(", ")}}` : "";
      return `${f.name}: ${f.typeName}${q}${c}`;
    };
    const lines: string[] = [];
    for (const s of this.structs) lines.push(`struct ${s.name} {${s.fields.map(fieldStr).join(", ")}}`);
    for (const f of this.fields) lines.push(fieldStr(f));
    for (const fn of this.functions) {
      const params = fn.params.map((p) => `${p.name}: ${p.typeName}${p.optional ? "?" : ""}`).join(", ");
      lines.push(`${fn.name}: (${params}) -> ${fn.returnType}`);
    }
    return lines.join("\n");
  }
}

function exportValue(v: unknown): string {
  if (typeof v === "string") return `'${v}'`;
  if (Array.isArray(v)) return `[${v.map(exportValue).join(", ")}]`;
  return String(v);
}

class SchemaParser {
  private pos = 0;
  private line = 1;

  constructor(private readonly text: string) {}

  private peek(): string | undefined {
    return this.text[this.pos];
  }

  private advance(): string {
    const ch = this.text[this.pos] as string;
    if (ch === "\n") this.line++;
    this.pos++;
    return ch;
  }

  private skipWS(newlines: boolean): void {
    while (this.pos < this.text.length) {
      const ch = this.text[this.pos];
      if (ch === "#") {
        while (this.pos < this.text.length && this.text[this.pos] !== "\n") this.pos++;
      } else if (ch === " " || ch === "\t") {
        this.pos++;
      } else if (newlines && (ch === "\n" || ch === "\r")) {
        if (ch === "\n") this.line++;
        this.pos++;
      } else {
        return;
      }
    }
  }

  private skipH(): void {
    while (this.text[this.pos] === " " || this.text[this.pos] === "\t") this.pos++;
  }

  private readIdent(): string {
    const m = IDENT.exec(this.text.slice(this.pos));
    if (!m) throw new AminoError("schema_parse", `expected identifier at line ${this.line}`);
    this.pos += m[0].length;
    return m[0];
  }

  private expect(ch: string): void {
    this.skipH();
    if (this.peek() !== ch) {
      throw new AminoError("schema_parse", `expected "${ch}" at line ${this.line}`);
    }
    this.advance();
  }

  private parseStrLiteral(): string {
    this.advance();
    const start = this.pos;
    while (this.peek() !== undefined && this.peek() !== "'") this.advance();
    if (this.peek() !== "'") throw new AminoError("schema_parse", `unterminated string at line ${this.line}`);
    const s = this.text.slice(start, this.pos);
    this.advance();
    return s;
  }

  private parseListLit(): unknown[] {
    this.advance();
    const items: unknown[] = [];
    this.skipH();
    while (this.peek() !== undefined && this.peek() !== "]") {
      items.push(this.parseConstraintVal());
      this.skipH();
      if (this.peek() === ",") {
        this.advance();
        this.skipH();
      }
    }
    if (this.peek() !== "]") throw new AminoError("schema_parse", `unterminated list at line ${this.line}`);
    this.advance();
    return items;
  }

  private parseConstraintVal(): unknown {
    this.skipH();
    const ch = this.peek();
    if (ch === "'") return this.parseStrLiteral();
    if (ch === "[") return this.parseListLit();
    const rest = this.text.slice(this.pos);
    let m = FLOAT.exec(rest);
    if (m) {
      this.pos += m[0].length;
      return parseFloat(m[0]);
    }
    m = INT.exec(rest);
    if (m) {
      this.pos += m[0].length;
      return parseInt(m[0], 10);
    }
    m = BOOL.exec(rest);
    if (m) {
      this.pos += m[0].length;
      return m[0] === "true";
    }
    throw new AminoError("schema_parse", `expected constraint value at line ${this.line}`);
  }

  private parseConstraints(): Record<string, unknown> {
    this.advance();
    const out: Record<string, unknown> = {};
    this.skipH();
    while (this.peek() !== "}") {
      if (this.peek() === undefined) throw new AminoError("schema_parse", `unterminated constraint block at line ${this.line}`);
      const key = this.readIdent();
      this.expect(":");
      out[key] = this.parseConstraintVal();
      this.skipH();
      if (this.peek() === ",") {
        this.advance();
        this.skipH();
      }
    }
    this.advance();
    return out;
  }

  private parseTypeExpr(): { kind: Kind; typeName: string; elementTypes: string[] } {
    this.skipH();
    const name = this.readIdent();
    if (name === "List") {
      this.expect("[");
      const elems = [this.readIdent()];
      while (this.peek() === "|") {
        this.advance();
        elems.push(this.readIdent());
      }
      this.expect("]");
      return { kind: "List", typeName: `List[${elems.join("|")}]`, elementTypes: elems };
    }
    const prim = PRIMITIVES[name];
    if (prim) return { kind: prim, typeName: name, elementTypes: [] };
    return { kind: "Custom", typeName: name, elementTypes: [] };
  }

  private parseField(): Field {
    this.skipH();
    const name = this.readIdent();
    if (RESERVED.has(name)) throw new AminoError("schema_parse", `reserved word "${name}" used as field name at line ${this.line}`);
    this.expect(":");
    const t = this.parseTypeExpr();
    const f: Field = { name, ...t, constraints: {}, optional: false };
    this.skipH();
    if (this.peek() === "?") {
      f.optional = true;
      this.advance();
    }
    this.skipH();
    if (this.peek() === "{") f.constraints = this.parseConstraints();
    return f;
  }

  private parseStruct(): StructDef {
    this.readIdent();
    this.skipH();
    const name = this.readIdent();
    this.skipWS(true);
    this.expect("{");
    const fields: Field[] = [];
    for (;;) {
      this.skipWS(true);
      if (this.peek() === undefined) throw new AminoError("schema_parse", `unterminated struct "${name}"`);
      if (this.peek() === "}") break;
      fields.push(this.parseField());
      this.skipH();
      if (this.peek() === ",") this.advance();
    }
    this.advance();
    return { name, fields };
  }

  private isFunction(): boolean {
    const saved = { pos: this.pos, line: this.line };
    try {
      this.readIdent();
      this.skipH();
      if (this.peek() !== ":") return false;
      this.advance();
      this.skipH();
      return this.peek() === "(";
    } catch {
      return false;
    } finally {
      this.pos = saved.pos;
      this.line = saved.line;
    }
  }

  private parseFunction(): FunctionDef {
    const name = this.readIdent();
    this.expect(":");
    this.expect("(");
    const params: Param[] = [];
    this.skipH();
    while (this.peek() !== ")") {
      if (this.peek() === undefined) throw new AminoError("schema_parse", `unterminated parameter list at line ${this.line}`);
      const pname = this.readIdent();
      this.expect(":");
      const { typeName } = this.parseTypeExpr();
      const param: Param = { name: pname, typeName, optional: false };
      this.skipH();
      if (this.peek() === "?") {
        param.optional = true;
        this.advance();
      }
      params.push(param);
      this.skipH();
      if (this.peek() === ",") {
        this.advance();
        this.skipH();
      }
    }
    this.advance();
    this.skipH();
    if (!this.text.startsWith("->", this.pos)) throw new AminoError("schema_parse", `expected "->" at line ${this.line}`);
    this.pos += 2;
    this.skipH();
    const returnType = this.readIdent();
    this.skipH();
    if (this.peek() === "?") this.advance();
    return { name, params, returnType };
  }

  parse(): Schema {
    const schema = new Schema();
    for (;;) {
      this.skipWS(true);
      if (this.pos >= this.text.length) break;
      if (STRUCT_KW.test(this.text.slice(this.pos))) schema.structs.push(this.parseStruct());
      else if (this.isFunction()) schema.functions.push(this.parseFunction());
      else schema.fields.push(this.parseField());
    }
    return schema;
  }
}

/** Parses .amn text. Validation happens in loadSchema, which knows the custom types. */
export function parseSchema(text: string): Schema {
  return new SchemaParser(text).parse();
}
